File: lab/ContourEvaluationMatrix_test/MultipleLoop.cc

#include "MultipleLoop.h"

MultipleLoop::MultipleLoop(int** loopLimits, int dim) {
	numberOfLoops = dim;

	limits = (int *) calloc(numberOfLoops, sizeof(int));
	pt1 = (int *) calloc(numberOfLoops, sizeof(int));

	loopsize = 1;

	for (int pont = 0; pont < numberOfLoops; pont++) {
		int* tmpPoint = loopLimits[pont];
		pt1[pont] = tmpPoint[0];
		int pt2 = tmpPoint[1];


		limits[pont] = pt2 - pt1[pont] + 1;
		loopsize *= limits[pont];
	}
}
  
MultipleLoop::~MultipleLoop(void) {
	free(limits);
	free(pt1);
}
    
int MultipleLoop::getLoopSize() {
	return loopsize;
}

int* MultipleLoop::getIndex(int pos) {
	int temp = pos;
	int *index = (int *) calloc(numberOfLoops, sizeof(int));

	for(int pont = 0; pont < numberOfLoops; pont++){
		index[pont] = (temp % limits[pont]) + pt1[pont];
		temp = temp / limits[pont];
	}

	return index;
}

File: desktop/src/java/wave/util/MultipleLoop.java
package wave.util;

public class MultipleLoop {

	private int numberOfLoops;
	private int limits[];
	private int pt1[];

	private int loopsize;

	public MultipleLoop(IntegerInterval[] looplimits) {
		numberOfLoops = looplimits.length;

		this.limits = new int[numberOfLoops];
		pt1 = new int[numberOfLoops];

		loopsize = 1;

		for (int pont = 0; pont < numberOfLoops; pont++) {
			pt1[pont] = looplimits[pont].getPt1();
			int pt2 = looplimits[pont].getPt2();


			limits[pont] = pt2 - pt1[pont] + 1;
			loopsize *= limits[pont];
		}
	}

	public int getLoopSize(){
		return loopsize;
	}

	public int[] getIndex(int pos) {
		int temp = pos;
		int index[] = new int[numberOfLoops];

		for(int pont = 0; pont < numberOfLoops; pont++){
			index[pont] = (temp % limits[pont]) + pt1[pont];
			temp = temp / limits[pont];
		}

		return index;
	}
}
